repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/opentracing/httpclient/ReposeTracingRequestInterceptor.scala

```diff
@@ -19,47 +19,77 @@
   *
   * @param tracer a [[io.opentracing.Tracer]] to bridge this utility with the OpenTracing API
   */
-class ReposeTracingRequestInterceptor(tracer: Tracer, reposeVersion: String, uriRedactionService: UriRedactionService)
-  extends SpanCreationRequestInterceptor(tracer) {
-
-  private val spanInjectionInterceptor = new SpanInjectionRequestInterceptor(tracer)
+class ReposeTracingRequestInterceptor(tracer: Tracer, reposeVersion: String, uriRedactionService: UriRedactionService) extends HttpRequestInterceptor with StrictLogging {
 
+  @throws[HttpException]
+  @throws[IOException]
   override def process(httpRequest: HttpRequest, httpContext: HttpContext): Unit = {
-    super.process(httpRequest, httpContext)
+    try {
+      val currentActiveSpan = tracer.activeSpan
+      val requestLine = httpRequest.getRequestLine
+      val redactedUri = uriRedactionService.redact(requestLine.getUri)
 
-    spanInjectionInterceptor.process(httpRequest, httpContext)
-  }
+      val clientSpanBuilder = tracer.buildSpan(s"${requestLine.getMethod} $redactedUri")
+      if (currentActiveSpan != null) clientSpanBuilder.asChildOf(currentActiveSpan)
 
-  override protected def onSpanStarted(clientSpan: Span, httpRequest: HttpRequest, httpContext: HttpContext): Unit = {
-    Option(httpRequest.getFirstHeader(REQUEST_ID))
-      .map(_.getValue)
-      .foreach(clientSpan.setTag(REQUEST_ID, _))
-    Option(httpRequest.getFirstHeader(VIA))
-      .map(_.getValue)
-      .foreach(clientSpan.setTag(VIA, _))
-    clientSpan.setTag(ReposeVersion, reposeVersion)
-    // Replace the http.url tag so that we do not leak the raw URL being set in this tag by the underlying Jaeger interceptor implementation
-    clientSpan.setTag(HTTP_URL.toString, uriRedactionService.redact(httpRequest.getRequestLine.getUri))
-  }
+      val clientSpan = clientSpanBuilder.start
+
+      clientSpan.setTag(SPAN_KIND.toString, SPAN_KIND_CLIENT)
+      clientSpan.setTag(HTTP_URL.toString, redactedUri)
+
+      httpContext match {
+        case context: HttpClientContext =>
+          val host = context.getTargetHost
+          clientSpan.setTag(PEER_HOSTNAME.toString, host.getHostName)
+          clientSpan.setTag(PEER_PORT.toString, host.getPort)
+        case _ =>
+      }
 
-  override protected def getOperationName(httpRequest: HttpRequest): String =
-    s"${httpRequest.getRequestLine.getMethod} ${uriRedactionService.redact(httpRequest.getRequestLine.getUri)}"
+      Option(httpRequest.getFirstHeader(REQUEST_ID))
+        .map(_.getValue)
+        .foreach(clientSpan.setTag(REQUEST_ID, _))
+      Option(httpRequest.getFirstHeader(VIA))
+        .map(_.getValue)
+        .foreach(clientSpan.setTag(VIA, _))
+      clientSpan.setTag(ReposeVersion, reposeVersion)
+
+      httpContext.setAttribute("io.opentracing.Span", clientSpan)
+
+      if (tracer.scopeManager.active == null) logger.warn("Current scope is null; possibly failed to start client tracing span.")
```

Comment:
Ugh. This is the case where Scala's `Option` always lets me down. I would like to take some action if an `Option` is `None`, but the ways we have to do that aren't the most elegant:
```
Option(tracer.scopeManager.active).getOrElse(logger.warn("Current scope is null; possibly failed to start client tracing span."))
```
```
if (Option(tracer.scopeManager.active).isEmpty) logger.warn("Current scope is null; possibly failed to start client tracing span.")
```
```
Option(tracer.scopeManager.active) match {
  case None => logger.warn("Current scope is null; possibly failed to start client tracing span.")
  case _ => // do nothing
}
```

All of which is debatably better than a straight null check.

Reply:
I know the `null` check is not the preferred way of doing this, but the only other option that seems to convey the same clarity is the second one. The first is performing the `get` for no reason and makes it fuzzy of what the intent is and the third is just so much extra for no gain. They become convention for convention's sake and make them harder to gain the actual intent.

Reply:
I agree, which is why I dislike the ways that Scala gives us to handle this scenario. I know other Scala folks have used `implicit`s to extend `Option` to provide an `ifNone` (or equivalent) method that invokes some function if the `Option` is `None`, which is something I've been tempted to do...

At any rate, not a big deal, do what seems best here.